dkc/core/rest/folder.py

```diff
@@ -147,10 +220,102 @@ def agree_terms(self, request, pk=None):
                 {'checksum': 'Mismatched checksum. Your terms may be out of date.'}
             )
 
-        # TODO get the actual user once REST authentication is in place
-        user = User.objects.first()
-
         TermsAgreement.objects.update_or_create(
-            terms=terms, user=user, defaults={'checksum': checksum}
+            terms=terms, user=request.user, defaults={'checksum': checksum}
         )
         return Response(status=204)
+
+    @swagger_auto_schema(
+        operation_description='Get all user or group permissions set on a folder',
+        responses={200: FolderPermissionGrantSerializer(many=True)},
+    )
+    @action(detail=True, methods=['get'], permission_classes=[IsAdmin])
+    def permissions(self, request, pk=None):
+        tree: Tree = self.get_object().tree
+        grants = tree.list_granted_permissions()
+        serializer = FolderPermissionGrantSerializer(grants, many=True)
+        return Response(serializer.data)
+
+    @transaction.atomic
+    def grant_permission_list(self, tree: Tree, grants: List[PermissionGrant]):
+        for grant in grants:
+            tree.grant_permission(grant)
+
+    @transaction.atomic
+    def remove_permission_list(self, tree: Tree, grants: List[PermissionGrant]):
+        for grant in grants:
+            tree.remove_permission(grant)
+
+    @transaction.atomic
+    def set_permission_list(self, tree: Tree, grants: List[PermissionGrant]):
+        self.grant_permission_list(tree, grants)
+
+        grant_set = set(grants)
+        existing_grant_set = set(tree.list_granted_permissions())
+        self.remove_permission_list(tree, list(existing_grant_set - grant_set))
```

Comment:
This explains the `frozen`-ness of `PermissionGrant`. I like the set-based approach, but does it make sense to move all this stuff (`set_permission_list`, `remove_permission_list`, `grant_permission_list`) down into the Tree model?